spec/unit/provider/git_spec.rb

```diff
@@ -590,6 +590,22 @@
     end
   end
 
+  context "why-run mode" do
```

Comment:
This actually missing why-run mode test cases. Recommend to update this.

Reply:
So what I observed is running this in why-run mode raises Argument Error 
```
Recipe: git_test::default
  * git[/tmp/mydir] action sync[2020-04-16T18:52:14+05:30] INFO: Processing git[/tmp/mydir] action sync (git_test::default line 7)

    * user_created_in_chef_run required by resource /tmp/mydir does not exist
    * User user_created_in_chef_run does not exist, this run will fail unless it has been previously created. Assuming it would have been created.
    ================================================================================
    Error executing action `sync` on resource 'git[/tmp/mydir]'
    ================================================================================
    
    ArgumentError
    -------------
    can't find user for user_created_in_chef_run
    
    Resource Declaration:
    ---------------------
    # In /home/msys/workspace/chef-repo-local/.chef/local-mode-cache/cache/cookbooks/git_test/recipes/default.rb
    
      7: git '/tmp/mydir' do
      8:   repository 'ssh://git@somehost.tld/myrepo.git'
      9:   revision 'master'
     10:   user 'user_created_in_chef_run'
     11:   group 'user_created_in_chef_run'
     12:   action :sync
     13: end
    
    Compiled Resource:
    ------------------
    # Declared in /home/msys/workspace/chef-repo-local/.chef/local-mode-cache/cache/cookbooks/git_test/recipes/default.rb:7:in `from_file'
    
    git("/tmp/mydir") do
      action [:sync]
      default_guard_interpreter :default
      declared_type :git
      cookbook_name "git_test"
      recipe_name "default"
      repository "ssh://git@somehost.tld/myrepo.git"
      revision "master"
      user "user_created_in_chef_run"
      group "user_created_in_chef_run"
      destination "/tmp/mydir"
      additional_remotes {}
    end
    
    System Info:
    ------------
    chef_version=16.0.191
    platform=ubuntu
    platform_version=18.04
    ruby=ruby 2.6.5p114 (2019-10-01 revision 67812) [x86_64-linux]
    program_name=/opt/chefdk/embedded/lib/ruby/gems/2.6.0/bin/chef-client
    executable=/opt/chefdk/embedded/bin/bundle
    
[2020-04-16T18:52:14+05:30] INFO: Running queued delayed notifications before re-raising exception

Running handlers:
[2020-04-16T18:52:14+05:30] ERROR: Running exception handlers
Running handlers complete
[2020-04-16T18:52:14+05:30] ERROR: Exception handlers complete
Chef Infra Client failed. 0 resources would have been updated
[2020-04-16T18:52:14+05:30] FATAL: Stacktrace dumped to /home/msys/workspace/chef-repo-local/.chef/local-mode-cache/cache/chef-stacktrace.out
[2020-04-16T18:52:14+05:30] FATAL: Please provide the contents of the stacktrace.out file if you file a bug report
[2020-04-16T18:52:14+05:30] FATAL: ArgumentError: git[/tmp/mydir] (git_test::default line 7) had an error: ArgumentError: can't find user for user_created_in_chef_run
```

And without why-run mode it raises 

```
Chef::Exceptions::User: git[/tmp/mydir] (git_test::default line 7) had an error: Chef::Exceptions::User: user_created_in_chef_run required by resource /tmp/mydir does not exist
```

So our test case should cover this both.

Reply:
So these are the test cases to go here. @tas50 need your review on this suggestion.

```
context "with why-run mode" do
    before do
      Chef::Config[:why_run] = true
      @resource.user "test"
    end

    after do
      Chef::Config[:why_run] = false
    end

    it "does not raise an error if user does not exist" do
      allow(@provider).to receive(:safe_get_homedir).with(@resource.user).and_return(nil)
      expect { @provider.run_action(:sync) }.not_to raise_error
    end

    it "Does not raise an error if user exists" do
      allow(@provider).to receive(:get_homedir).with(@resource.user).and_return("/home/test")
      expect { @provider.run_action(:sync) }.not_to raise_error(ArgumentError)
    end
  end

  context "without why-run mode" do
    before do
      @resource.user "test"
    end

    it "raises an error if user does not exist" do
      allow(@provider).to receive(:safe_get_homedir).with(@resource.user).and_return(nil)
      expect { @provider.run_action(:sync) }.to raise_error(Chef::Exceptions::User)
    end

    it "Does not raise an error if user exists" do
      allow(@provider).to receive(:get_homedir).with(@resource.user).and_return("/home/test")
      expect { @provider.run_action(:sync) }.not_to raise_error(Chef::Exceptions::User)
    end
  end
```

Reply:
That seems to be our standard pattern